File: apps/web/src/services/manufacturingMrp.ts
import { apiClient } from './apiClient';

export interface BOMItem {
  id?: string;
  raw_material_id: string;
  quantity: number;
  unit: string;
  scrap_percentage: number;
  sequence: number;
}

export interface BillOfMaterial {
  id: string;
  product_id: string;
  version_id?: string;
  bom_code: string;
  revision: string;
  description?: string;
  status: string;
  created_at: string;
  items: BOMItem[];
}

export interface WorkCenter {
  id: string;
  organization_id: string;
  center_name: string;
  center_code: string;
  capacity: number;
  location?: string;
  status: string;
  created_at: string;
}

export interface Machine {
  id: string;
  work_center_id: string;
  machine_name: string;
  machine_code: string;
  manufacturer?: string;
  serial_number?: string;
  status: string;
  created_at: string;
}

export interface ProductionOrder {
  id: string;
  organization_id: string;
  production_number: string;
  product_id: string;
  planned_quantity: number;
  completed_quantity: number;
  scheduled_start: string;
  scheduled_end: string;
  priority: string;
  status: string;
  created_at: string;
}

export interface QualityInspection {
  id: string;
  production_order_id: string;
  inspection_type: string;
  inspection_date: string;
  status: string;
  created_at: string;
}

export interface MRPRecommendation {
  product_id: string;
  product_name: string;
  required_quantity: number;
  current_stock: number;
  shortage_quantity: number;
  action_type: string;
}

export interface MRPRunResponse {
  id: string;
  organization_id: string;
  run_date: string;
  planning_period: string;
  status: string;
  processed_items: number;
  recommendations: MRPRecommendation[];
  created_at: string;
}

export interface ManufacturingDashboardSummary {
  active_production_orders: number;
  machine_utilization_rate: number;
  total_material_consumed: number;
  production_efficiency_percentage: number;
  total_production_cost: number;
  quality_pass_rate_percentage: number;
  mrp_recommendations_count: number;
  maintenance_schedules_count: number;
}

export const manufacturingMrpService = {
  // Dashboard
  getDashboardSummary: async (orgId: string): Promise<ManufacturingDashboardSummary> => {
    const res = await apiClient.get('/api/v1/manufacturing/dashboard', { params: { org_id: orgId } });
    return res.data;
  },

  // BOM
  getBOMs: async (): Promise<BillOfMaterial[]> => {
    const res = await apiClient.get('/api/v1/manufacturing/bom');
    return res.data;
  },

  createBOM: async (data: any): Promise<BillOfMaterial> => {
    const res = await apiClient.post('/api/v1/manufacturing/bom', data);
    return res.data;
  },

  // Production Orders
  getProductionOrders: async (orgId: string): Promise<ProductionOrder[]> => {
    const res = await apiClient.get('/api/v1/manufacturing/production-orders', { params: { org_id: orgId } });
    return res.data;
  },

  createProductionOrder: async (data: any): Promise<ProductionOrder> => {
    const res = await apiClient.post('/api/v1/manufacturing/production-orders', data);
    return res.data;
  },

  startProductionOrder: async (id: string): Promise<ProductionOrder> => {
    const res = await apiClient.post(`/api/v1/manufacturing/production-orders/${id}/start`);
    return res.data;
  },

  completeProductionOrder: async (id: string): Promise<ProductionOrder> => {
    const res = await apiClient.post(`/api/v1/manufacturing/production-orders/${id}/complete`);
    return res.data;
  },

  // Work Centers & Machines
  getWorkCenters: async (orgId: string): Promise<WorkCenter[]> => {
    const res = await apiClient.get('/api/v1/manufacturing/work-centers', { params: { org_id: orgId } });
    return res.data;
  },

  getMachines: async (): Promise<Machine[]> => {
    const res = await apiClient.get('/api/v1/manufacturing/machines');
    return res.data;
  },

  // MRP
  runMRP: async (orgId: string, period: string): Promise<MRPRunResponse> => {
    const res = await apiClient.post('/api/v1/manufacturing/mrp/run', {
      organization_id: orgId,
      planning_period: period,
    });
    return res.data;
  },

  // Quality
  getQualityInspections: async (): Promise<QualityInspection[]> => {
    const res = await apiClient.get('/api/v1/manufacturing/quality');
    return res.data;
  },
};

File: apps/api/app/schemas/manufacturing_mrp.py
import uuid
from datetime import date, datetime
from pydantic import BaseModel, ConfigDict, Field


# --- BOM Schemas ---
class BOMItemBase(BaseModel):
    raw_material_id: uuid.UUID
    quantity: float = Field(default=1.0, gt=0)
    unit: str = Field(default="Pcs", max_length=50)
    scrap_percentage: float = Field(default=0.0, ge=0)
    sequence: int = Field(default=1, ge=1)


class BOMItemCreate(BOMItemBase):
    pass


class BOMItemResponse(BOMItemBase):
    id: uuid.UUID
    bom_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BOMBase(BaseModel):
    bom_code: str = Field(..., min_length=1, max_length=100)
    revision: str = Field(default="Rev A", max_length=50)
    description: str | None = Field(None, max_length=1000)
    status: str = Field(default="Active", max_length=50)


class BOMCreate(BOMBase):
    product_id: uuid.UUID
    version_id: uuid.UUID | None = None
    items: list[BOMItemCreate] = Field(..., min_length=1)


class BOMResponse(BOMBase):
    id: uuid.UUID
    product_id: uuid.UUID
    version_id: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime
    items: list[BOMItemResponse] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


# --- Work Center & Machine Schemas ---
class WorkCenterBase(BaseModel):
    center_name: str = Field(..., min_length=1, max_length=255)
    center_code: str = Field(..., min_length=1, max_length=50)
    capacity: float = Field(default=100.0, gt=0)
    location: str | None = Field(None, max_length=255)
    status: str = Field(default="Active", max_length=50)


class WorkCenterCreate(WorkCenterBase):
    organization_id: uuid.UUID
    manager_uuid: uuid.UUID | None = None


class WorkCenterResponse(WorkCenterBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    manager_uuid: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class MachineBase(BaseModel):
    machine_name: str = Field(..., min_length=1, max_length=255)
    machine_code: str = Field(..., min_length=1, max_length=50)
    manufacturer: str | None = Field(None, max_length=255)
    serial_number: str | None = Field(None, max_length=100)
    installation_date: date | None = None
    status: str = Field(default="Operational", max_length=50)


class MachineCreate(MachineBase):
    work_center_id: uuid.UUID


class MachineResponse(MachineBase):
    id: uuid.UUID
    work_center_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class MachineMaintenanceBase(BaseModel):
    maintenance_type: str = Field(default="Preventive", max_length=50)
    scheduled_date: date
    completed_date: date | None = None
    cost: float = Field(default=0.0, ge=0)
    status: str = Field(default="Scheduled", max_length=50)


class MachineMaintenanceCreate(MachineMaintenanceBase):
    machine_id: uuid.UUID


class MachineMaintenanceResponse(MachineMaintenanceBase):
    id: uuid.UUID
    machine_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Production Order Schemas ---
class ProductionOrderBase(BaseModel):
    production_number: str = Field(..., min_length=1, max_length=100)
    planned_quantity: float = Field(..., gt=0)
    scheduled_start: date
    scheduled_end: date
    priority: str = Field(default="Medium", max_length=50)
    status: str = Field(default="Draft", max_length=50)


class ProductionOrderCreate(ProductionOrderBase):
    organization_id: uuid.UUID
    product_id: uuid.UUID


class ProductionOrderResponse(ProductionOrderBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    product_id: uuid.UUID
    completed_quantity: float
    actual_start: date | None = None
    actual_end: date | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Quality Inspection Schemas ---
class QualityInspectionBase(BaseModel):
    inspection_type: str = Field(default="In-Process", max_length=50)
    inspection_date: date
    status: str = Field(default="Passed", max_length=50)


class QualityInspectionCreate(QualityInspectionBase):
    production_order_id: uuid.UUID
    inspector_id: uuid.UUID | None = None


class QualityInspectionResponse(QualityInspectionBase):
    id: uuid.UUID
    production_order_id: uuid.UUID
    inspector_id: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- MRP Schemas ---
class MRPRunCreate(BaseModel):
    organization_id: uuid.UUID
    planning_period: str = Field(..., min_length=1, max_length=100)


class MRPRecommendation(BaseModel):
    product_id: uuid.UUID
    product_name: str
    required_quantity: float
    current_stock: float
    shortage_quantity: float
    action_type: str  # Purchase Requisition, Production Order


class MRPRunResponse(BaseModel):
    id: uuid.UUID
    organization_id: uuid.UUID
    run_date: date
    planning_period: str
    status: str
    processed_items: int
    recommendations: list[MRPRecommendation] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Manufacturing Dashboard Summary ---
class ManufacturingDashboardSummary(BaseModel):
    active_production_orders: int
    machine_utilization_rate: float
    total_material_consumed: float
    production_efficiency_percentage: float
    total_production_cost: float
    quality_pass_rate_percentage: float
    mrp_recommendations_count: int
    maintenance_schedules_count: int
